core/src/main/java/hudson/model/AbstractBuild.java

```diff
@@ -459,49 +460,64 @@ public Result run(@Nonnull BuildListener listener) throws Exception {
+                            for (LabelAtom label : assignedLabels) {
+                                if (first) {
+                                    listener.getLogger().print(" (");
+                                    first = false;
+                                } else {
+                                    listener.getLogger().print(' ');
+                                }
+                                listener.getLogger().print(label.getName());
                             }
-                            listener.getLogger().print(label.getName());
+                            listener.getLogger().print(')');
                         }
-                        listener.getLogger().print(')');
                     }
+                } else {
+                    listener.getLogger().print(Messages.AbstractBuild_Building());
                 }
-            } else {
-                listener.getLogger().print(Messages.AbstractBuild_Building());
-            }
-            
-            lease = decideWorkspace(node, Computer.currentComputer().getWorkspaceList());
+                
+                lease = decideWorkspace(node, Computer.currentComputer().getWorkspaceList());
 
-            workspace = lease.path.getRemote();
-            listener.getLogger().println(Messages.AbstractBuild_BuildingInWorkspace(workspace));
-            node.getFileSystemProvisioner().prepareWorkspace(AbstractBuild.this,lease.path,listener);
+                workspace = lease.path.getRemote();
+                listener.getLogger().println(Messages.AbstractBuild_BuildingInWorkspace(workspace));
+                node.getFileSystemProvisioner().prepareWorkspace(AbstractBuild.this,lease.path,listener);
 
-            for (WorkspaceListener wl : WorkspaceListener.all()) {
-                wl.beforeUse(AbstractBuild.this, lease.path, listener);
-            }
+                for (WorkspaceListener wl : WorkspaceListener.all()) {
+                    wl.beforeUse(AbstractBuild.this, lease.path, listener);
+                }
 
-            getProject().getScmCheckoutStrategy().preCheckout(AbstractBuild.this, launcher, this.listener);
-            getProject().getScmCheckoutStrategy().checkout(this);
+                getProject().getScmCheckoutStrategy().preCheckout(AbstractBuild.this, launcher, this.listener);
+                getProject().getScmCheckoutStrategy().checkout(this);
 
-            if (!preBuild(listener,project.getProperties()))
-                return Result.FAILURE;
+                if (!preBuild(listener,project.getProperties()))
+                    return Result.FAILURE;
 
-            Result result = doRun(listener);
+                result = doRun(listener);
+            } finally {
+                if (!tearDownMarker.tornDown) {
+                    // looks like environments are not torn down yet, do it now (might affect the build result)
+                    result = Result.combine(result, tearDownBuildEnvironments(listener));
```

Comment:
InterruptedException in `tearDownBuildEnvironments()` will bypass this logic and cause loss of the original build result information. I doubt this is what we want

Reply:
That's intentional, see https://github.com/jenkinsci/jenkins/pull/4517#discussion_r385416603
My understanding of how it works today is that an `InterruptedException` during an environment `tearDown()` would go up to `Run#execute(RunExecution)`, where it is handled in a "special" way (`Executor#abortResult()`) to take user interruption cause into account. In the end, the result is `ABORTED`, effectively bypassing the result of the build steps.
I don't really want to change/rewrite that, it's not what this pull request is about. I've tried to change things "just enough" to make sure all environments are torn down even in this case (I'll still try to tear down remaining environments before re-throwing the exception), but no more than that.

I've added a test case (83c2602) which shows that:
1. the result is `ABORTED` in this case,
2. remaining environments are torn down.

Executing the same test without the rest of this pull-request still passes 1 (the `ABORTED` assertion), and fails on 2. (see thomasgl-orange/jenkins@dc71032fd7 for a cherry pick of 83c2602 on top of master).